### ISSUE TITLE
Error with CUDA for quantization aware training

### ISSUE THREAD

**Issue by pythorghst** [pytorch/pytorch#36802]
## 🐛 Bug

I'm following this tutorial : https://pytorch.org/tutorials/advanced/static_quantization_tutorial.html
This tutorial uses a CPU, it works fine with a CPU, but to speed it up I want to move it to the GPU and I get an error. It works with Pytorch version 1.4 but not with Pytorch nightly (1.5.0.dev20200413).

## To Reproduce

1.Follow the following tutorial:
https://pytorch.org/tutorials/advanced/static_quantization_tutorial.html

2. In the "5. Quantization-aware training" section

Instead of doing
```
for nepoch in range(8):
    train_one_epoch(qat_model, criterion, optimizer, data_loader, torch.device('cpu'), num_train_batches)
```
do :
```
device = torch.device("cuda:0")
for nepoch in range(8):
    device = torch.device("cuda:0")
    train_one_epoch(qat_model, criterion, optimizer, train_loader, torch.device(device), num_train_batches)
```
also changed in the section "2. Helper functions":
```
def load_model(model_file):
    model = MobileNetV2()
    state_dict = torch.load(model_file)
    model.load_state_dict(state_dict)
    model.to(device) #device instead of 'cpu'
    return model
```
3.Get the error
```console
F:\Anaconda3\envs\pytorch-nightly\lib\site-packages\torch\quantization\fake_quantize.py in forward(self, X)
     84             if self.qscheme == torch.per_channel_symmetric or self.qscheme == torch.per_channel_affine:
     85                 X = torch.fake_quantize_per_channel_affine(X, self.scale, self.zero_point,
---> 86                                                            self.ch_axis, self.quant_min, self.quant_max)
     87             else:
     88                 X = torch.fake_quantize_per_tensor_affine(X, float(self.scale),

RuntimeError: iter.device(arg).is_cuda() INTERNAL ASSERT FAILED at C:/cb/pytorch_1000000000000/work/aten/src\ATen/native/cuda/Loops.cuh:60, please report a bug to PyTorch. 
```
## Expected behavior

No error, run smoothly. QAT is supposed to support CUDA.
Quote from the documentation
"Quantization-aware training (through FakeQuantize) supports both CPU and CUDA."
 
However, it works with  PyTorch Version : 1.4.0.

## Environment

 - PyTorch Version : 1.5.0.dev20200413 (pytorch nightly)
 - OS :Windows 10
 - How you installed PyTorch :conda (version 4.7.12)
 - Build command you used (if compiling from source):
 conda install pytorch torchvision cudatoolkit=10.1 -c pytorch-nightly -c defaults -c conda-forge
 - Python version: 3.7.7
 - CUDA/cuDNN version: 10.1
 - GPU models and configuration: GPU: Nvidia Geforce RTX 2060 Super (8Go GDDR6)
								
Versions of relevant libraries:
[pip] numpy==1.18.1
[pip] torch==1.5.0.dev20200413
[pip] torchvision==0.7.0.dev20200415
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               10.1.243             h74a9793_0  
[conda] mkl                       2020.0                      166  
[conda] mkl-service               2.3.0            py37hb782905_0  
[conda] mkl_fft                   1.0.15           py37h14836fe_0  
[conda] mkl_random                1.1.0            py37h675688f_0  
[conda] numpy                     1.18.1           py37h93ca92e_0  
[conda] numpy-base                1.18.1           py37hc3f5095_1  
[conda] pytorch                   1.5.0.dev20200413 py3.7_cuda101_cudnn7_0    pytorch-nightly
[conda] torchvision               0.7.0.dev20200415      py37_cu101    pytorch-nightly

cc @jerryzh168 @jianyuh @dzhulgakov @raghuramank100 @jamesr66a

**Comment by peterjc123** [comment]
Yes, I could reproduce this issue. The workground is to do `qat_model.to(device='cuda')` again before the loop. Maybe some tensors are not transferred to GPU after calling `torch.quantization.prepare_qat`.

**Comment by pythorghst** [comment]
Thank you. It works with qat_model.to(device='cuda'). But it would be better not having to do this workaround. Plus, QAT is at least 10 times faster in 1.5 so a proper fix would be appreciated (or maybe it's the 1.4 version which is abnormaly slow)

**Comment by semchan** [comment]
hi, when I  run on DPU, pop the below error: 

code:
qmodel.eval().cuda()
qdummy = torch.quantize_per_tensor(dummy, scale=1, zero_point=0, dtype=torch.quint8).cuda() 

error:
Exception has occurred: RuntimeError
Copying from quantized Tensor to non-quantized Tensor is not allowed, please use dequantize to get a float Tensor from a quantized Tensor

can you please help me? thanks

**Comment by jerryzh168** [comment]
> hi, when I run on DPU, pop the below error:
> 
> code:
> qmodel.eval().cuda()
> qdummy = torch.quantize_per_tensor(dummy, scale=1, zero_point=0, dtype=torch.quint8).cuda()
> 
> error:
> Exception has occurred: RuntimeError
> Copying from quantized Tensor to non-quantized Tensor is not allowed, please use dequantize to get a float Tensor from a quantized Tensor
> 
> can you please help me? thanks

looks like this is a question, could you ask in https://discuss.pytorch.org/c/quantization/17

**Comment by vkuzo** [comment]
> Thank you. It works with qat_model.to(device='cuda'). But it would be better not having to do this workaround. Plus, QAT is at least 10 times faster in 1.5 so a proper fix would be appreciated (or maybe it's the 1.4 version which is abnormaly slow)

thanks for the report! https://github.com/pytorch/pytorch/pull/39031 should fix this, which is a part of v1.6.

**Comment by vkuzo** [comment]
closing this since we fixed all of these for v1.6.  @semchan , could you please let us know if you are still hitting these issues after updating to v1.6?

**Comment by Pranav-India** [comment]
Hi @vkuzo  I am having the following error ,I am using PyTorch version 1.7.0 and cuda 10.1.... can you please help me with this?



~/anaconda3/envs/pytorch_p36/lib/python3.6/site-packages/torch/nn/functional.py in mse_loss(input, target, size_average, reduce, reduction)
   2658 
   2659     expanded_input, expanded_target = torch.broadcast_tensors(input, target)
-> 2660     return torch._C._nn.mse_loss(expanded_input, expanded_target, _Reduction.get_enum(reduction))
   2661 
   2662 

RuntimeError: iter.device(arg).is_cuda() INTERNAL ASSERT FAILED at "/opt/conda/conda-bld/pytorch_1603728993639/work/aten/src/ATen/native/cuda/Loops.cuh":94, please report a bug to PyTorch.
